packages/dds/tree/src/core/tree/anchorSet.ts

```diff
@@ -668,6 +706,17 @@ export class AnchorSet implements Listenable<AnchorSetRootEvents>, AnchorLocator
 			pathVisitors: new Map<PathNode, Set<PathVisitor>>(),
 			parentField: undefined as FieldKey | undefined,
 			parent: undefined as UpPath | undefined,
+			bufferedEvents: [] as { node: PathNode; event: keyof AnchorEvents }[],
+
+			// A stack of booleans to keep track of when we should be enqueuing treeChanged events during tree traversal
+			// as a delta visit drives this visitor.
+			// As we go deeper into the tree we add entries to the stack, defaulting to false, and when a node actually changes
+			// we set the top of the stack to true.
+			// As we exit nodes, we pop the stack and if the popped value is true, that's an indication that we should enqueue
+			// a treeChanged event; it's also an indication that we should enqueue treeChanged events for all ancestors of the
+			// node, so we need to also update the top of the stack to ensure that keeps happening as we go up the tree.
+			actualChangeStack: [] as boolean[],
```

Comment:
Is it possible to do this tracking with a single number? 

Initialize _d_ to 0.
* When you enter a node, make no change to _d_.
* When you edit a node, set _d_ to your current depth.
* When you exit a node: iff _d_ == your current depth, emit a `subtreeChangedAfterBatch` and decrement _d_.

Intuitively, _d_ is the number of nodes, from the root of your current spine, that need to eventually fire the event.

Does that work? I'm not ultra confident but it sounds like it might work. 
I guess you'd need a second number to track the current depth, since we don't have that lying around right now, but it'd still be constant space rather than O(depth).

Reply:
Does seem to work, nice optimization! Thanks!

Reply:
Great! 

Alex P also mentioned that, in theory, we shouldn't need this state at all if the delta walk is "minimal" - i.e., the visitor only walks down to the nodes that changed, and no others. Because in that case, you could simply always emit `subtreeChangedAfterBatch` unconditionally every time you exited a node; all nodes in the walk have changed. I'm assuming that since we need this state here, the delta walk must not be minimal. However, Alex doesn't see any reason why we couldn't change it to do the minimal walk in the future.

Reply:
I know @yann-achard-MS had already thought about it and had ideas to implement a visitor wrapper (or similar) that did exactly that, skipped unnecessary parts of the walk in a given pass. We might need to touch a node during the attach pass but not the detach one, and if we don't walk to it during the detach pass, indeed this extra state becomes unnecessary and we could unconditionally emit the events. In that case it might even be possible to remove these new ones and do everything in terms of subtreeChanged and childrenChanged (still deduping, to handle the fact that they could be emitted for a node during both passes).